### DISCUSSION TITLE
Large Performance Delta between debug and release mode

### DISCUSSION THREAD

**Discussion by JustSomeRandomUsername** [image-rs/image#2371] [General]
When loading a simple image and encoding it with webp (lossless) my laptop takes:
|mode|load|encode|
|---|---|---|
|debug|105 ms|170 ms|
|release|5 ms|3 ms|

Image: 640x480 jpeg 12.5kB

I am quite surprised about the large delta. Is this something your aware of?
Do you know what is causing the slowdown ?

## Reproduction steps

```let t = Instant::now();
    let img = image::open("0.jpg").unwrap();
    let t2 = Instant::now();
    let mut output = vec![];
    let encoder = WebPEncoder::new_lossless(&mut output);
    img.write_with_encoder(encoder).unwrap();
    println!("Load {:?}, Encode {:?}", t2.duration_since(t), Instant::now().duration_since(t2));
```

**Comment by 197g** [comment]
(This may be a good point to move this to a discussion.)

Yes. It is quite expected. That's just how Rust compilation works.

You can change your `Cargo.toml` to _selectively_ choose another optimization level for dependencies. For instance, from a configuration in some of my own projects:

```toml
[profile.dev.package]
adler2.opt-level = 3
miniz_oxide.opt-level = 3
png.opt-level = 3
flate2.opt-level = 3
image.opt-level = 3
```

Keep in mind these are local to the workspace and the exact set will require a bit of fine-tuning. The level is also not recursive, I'd recommend petitioning with `cargo` if this presents a distinct problem. They'll probably love to hear _concrete_ use cases that would help design concepts to be validated effectively.

**Comment by JustSomeRandomUsername** [comment]
First of all thank you for the quick answer and sorry for not putting this in an discussion post directly. 

I played around a bit with the opt-levels.

```jpeg-decoder.opt-level = 3```  brings the decoding/loading times on par with release mode. That already great.

But even though I specified the opt-level of every  dependency of ```image-rs``` to be 3. I can't drop the WebP encoding times.
The only thing that seems to work is setting the opt-level of the whole package to 3, but that is not ideal in my case.

**Reply by fintelia** [reply]
I would have expected `image-webp.opt-level = 3` to be sufficient for that. Are you using image 0.24.x or 0.25.x?

**Reply by 197g** [reply]
Also, for a list of all dependency crates enabled in your workspace: `cargo tree -p image`

**Reply by JustSomeRandomUsername** [reply]
I have set 77 dependencies to opt-level=3 and it doesn't seem to change anything.
Running  `cargo tree -p image` gave me the list of the 77 i set. I hope I didn't loose one while removing the duplicates.

I see the same behaviour on 0.25.4 and 0.24.9.

**Reply by fintelia** [reply]
I wonder if `image-webp`'s use of generics is causing it not to be optimized
